src/devices/StUsb4500/samples/StUsb4500.Sample.cs

```diff
@@ -0,0 +1,180 @@
+
+            Console.WriteLine();
+        }
+
+        /// <summary>Prints the NVM data.</summary>
+        /// <param name="stUsb">The device.</param>
+        private static void PrintNvmData(StUsb4500 stUsb)
+        {
+            Console.WriteLine("NVM data:");
+            byte[] nvmData = stUsb.NvmData;
+            Console.WriteLine($"0x00: 0x{nvmData[0]:X2} 0x{nvmData[1]:X2} 0x{nvmData[2]:X2} 0x{nvmData[3]:X2} 0x{nvmData[4]:X2} 0x{nvmData[5]:X2} 0x{nvmData[6]:X2} 0x{nvmData[7]:X2}");
+            Console.WriteLine($"0x08: 0x{nvmData[8]:X2} 0x{nvmData[9]:X2} 0x{nvmData[10]:X2} 0x{nvmData[11]:X2} 0x{nvmData[12]:X2} 0x{nvmData[13]:X2} 0x{nvmData[14]:X2} 0x{nvmData[15]:X2}");
+            Console.WriteLine($"0x16: 0x{nvmData[16]:X2} 0x{nvmData[17]:X2} 0x{nvmData[18]:X2} 0x{nvmData[19]:X2} 0x{nvmData[20]:X2} 0x{nvmData[21]:X2} 0x{nvmData[22]:X2} 0x{nvmData[23]:X2}");
+            Console.WriteLine($"0x24: 0x{nvmData[24]:X2} 0x{nvmData[25]:X2} 0x{nvmData[26]:X2} 0x{nvmData[27]:X2} 0x{nvmData[28]:X2} 0x{nvmData[29]:X2} 0x{nvmData[30]:X2} 0x{nvmData[31]:X2}");
+            Console.WriteLine($"0x32: 0x{nvmData[32]:X2} 0x{nvmData[33]:X2} 0x{nvmData[34]:X2} 0x{nvmData[35]:X2} 0x{nvmData[36]:X2} 0x{nvmData[37]:X2} 0x{nvmData[38]:X2} 0x{nvmData[39]:X2}{Environment.NewLine}");
+        }
+
+        /// <summary>Updates the NVM data.</summary>
+        /// <param name="stUsb">The device.</param>
+        private static void UpdateNvmData(StUsb4500 stUsb)
+        {
+            byte[] nvmData =
+                {
+                    0x00, 0x00, 0xB0, 0xAA, 0x00, 0x45, 0x00, 0x00,
+                    0x10, 0x40, 0x9C, 0x1C, 0xF0, 0x01, 0x00, 0xDF,
+                    0x02, 0x40, 0x0F, 0x00, 0x32, 0x00, 0xFC, 0xF1,
+                    0x00, 0x19, 0x57, 0xAF, 0xF6, 0x4F, 0xFF, 0x00,
+                    0x00, 0x4B, 0x90, 0x21, 0x03, 0x00, 0x48, 0xFB,
+                };
+            stUsb.NvmData = nvmData;
+            PrintNvmData(stUsb);
+        }
+
+        /// <summary>Prints the source PDOs.</summary>
+        /// <param name="stUsb">The device.</param>
+        private static void PrintSourcePdo(StUsb4500 stUsb)
+        {
+            Power maxPower = Power.Zero;
+            Console.WriteLine("Source PDOs:");
+            PowerDeliveryObject[] sourcePdos = stUsb.SourcePowerDeliveryObjects;
+            for (int i = 0; i < sourcePdos.Length; i++)
+            {
+                PowerDeliveryObject pdo = sourcePdos[i];
+                Console.WriteLine($"PDO #{i + 1}: {pdo}");
+                if (pdo.Power > maxPower)
+                {
+                    maxPower = pdo.Power;
+                }
+            }
+
+            Console.WriteLine($"P(max) = {maxPower:0.##}{Environment.NewLine}");
+        }
+
+        /// <summary>Prints the RDO.</summary>
+        /// <param name="stUsb">The device.</param>
+        private static void PrintRdo(StUsb4500 stUsb)
+        {
+            Console.WriteLine("RDO (negotiated power):");
```

Comment:
please avoid using ToString in the sample, most likely users will want to know how to access the properties and printing them while informational is not useful in the real app. If the code seems repetitive when accessing properties you should consider changing APIs so it's not.

Reply:
@krwq I need to disagree as these values are useful to be printed in a real app - this is exactly what we do and is the reason I created this device binding. We show some of these values in our web-interface to inform the user of our device if a valid USB-PD source is connected (as USB-Pd is only one of several ways to power our device). It is also a copy of the original example from the official STUSB4500 repo (https://github.com/usb-c/STUSB4500/tree/master/Firmware).